README.md
# Instructions:

- create symbolic link to vimrc `ln -s ~/.vimrc ~/.vim/vimrc`
- create symbolic link to gvimrc `ln -s ~/.gvimrc ~/.vim/gvimrc`
- install jshint
- create symbolic link to jshintrc `ln -s ~/.jshintrc ~/.vim/jshintrc`
- install exuberant ctags
- install jsctags  


docutils/HISTORY.txt
.. -*- coding: utf-8 -*-

==================
 Docutils History
==================

:Author: David Goodger; open to all Docutils developers
:Contact: goodger@python.org
:Date: $Date: 2012-05-02 19:13:41 +0200 (Mi, 02 Mai 2012) $
:Revision: $Revision: 7410 $
:Web site: http://docutils.sourceforge.net/
:Copyright: This document has been placed in the public domain.

.. contents::

Release 0.9 (2012-05-02)
========================

* General:

  - New reStructuredText "code" role and directive and "code" option
    of the "include" directive with syntax highlighting by Pygments_.
  - Fix parse_option_marker for option arguments containing ``=``.

.. _Pygments: http://pygments.org/

* setup.py

  - Fix [ 2971827 ] and [ 3442827 ]
    extras/roman.py moved to docutils/utils/roman.py

* docutils/frontend.py

  - Fix [ 3481980 ] Use os.getcwdu() in make_paths_absolute().

* docutils/io.py

  - Fix [ 3395948 ] (Work around encoding problems in Py3k).
  - `mode` argument for FileOutput avoids code replication in
    BinaryFileOutput.
  - New exceptions InputError and OutputError for IO errors in
    FileInput/FileOutput.

* docutils/core.py:

  - No "hard" system exit on file IO errors: catch and report them in
    `Publisher.reportException` instead. Allows handling by a calling
    application if the configuration setting `traceback` is True.

* docutils/utils.py -> docutils/utils/__init__.py

  - docutils.utils is now a package (providing a place for sub-modules)

  .. note:: docutils/math, docutils/error_reporting.py, and
     docutils/urischemes.py will move to the utils package in the next
     release, too. See RELEASE-NOTES__

     __ RELEASE-NOTES.html

  - DependencyList uses io.FileOutput and 'utf8' encoding to prevent
    errors recording non-ASCII filenames (fixes [ 3434355 ]).

  - Fix relative_path() with source=None and `unicode` target.

* docutils/parsers/rst/states.py

  - Fix [ 3402314 ] allow non-ASCII whitespace, punctuation
    characters and "international" quotes around inline markup.
  - Use `field_marker` pattern to look for start of a
    directive option block (fixes [ 3484857 ]).

* docutils/parsers/rst/tableparser.py

  - Fix [ 2926161 ] for simple tables.
    (Combining chars in grid tables still contribute to cell width.)

* docutils/writers/latex2e/__init__.py

  - Support the `abbreviation` and `acronym` standard roles.
  - Record only files required to generate the LaTeX source as dependencies.
  - Fix handling of missing stylesheets.
  - Use ``\setcounter{secnumdepth}{0}`` instead of ``*``-versions
    when suppressing LaTeX section numbering.
  - Use ``\DUtitle`` for unsupported section levels
  - Apply [ 3512791 ] do not compare string literals with "is"

* docutils/writers/xetex/__init__.py

  - Avoid code duplication with latex2e writer (solves [ 3512728 ]).

* docutils/writers/html4css1/__init__.py

  - Change default for `math-output` setting to MathJax.
  - Fix handling of missing stylesheets.

* docutils/writers/docutils_xml.py

  - Use the visitor pattern with default_visit()/default_depart() methods
    instead of minidom to facilitate special handling of selected nodes.
  - Support raw XML (inserted as-is inside a <raw></raw> node).

* docutils/writers/manpage.py

  - Do not emit comment line with trailing blank. Problematic for VCS.

Release 0.8.1 (2011-08-30)
==========================

* General:

  - Fix [ 3364658 ] (Change last file with Apache license to BSD-2-Clause)
    and [ 3395920 ] (correct copyright info for rst.el).

* docutils/test/

  -  Apply [ 3303733 ] and [ 3365041 ] to fix tests under py3k.

* docutils/writers/latex2e/__init__.py

  - Clean up Babel language setting. Restores Sphinx compatibility.

Release 0.8 (2011-07-07)
========================

* General:

  - Handle language codes according to `BCP 47`_.
  - If the specified language is not supported by Docutils,
    warn and fall back to English.
  - Math support: reStructuredText "math" role and directive,
    ``math`` and ``math_block`` doctree elements.
  - Decode command line arguments with the locale's preferred encoding
    (to allow, e.g., ``--title=Dornröschen``).
  - Orphaned "python" reader and "newlatex2e" writer moved to the sandbox.
  - New sub-module `error_reporting`: handle encoding/decoding errors
    when reporting exceptions.
  - Some additions to the Docutils core are released under the 2-Clause BSD
    license, see COPYING_ for details.

  .. _BCP 47: http://www.rfc-editor.org/rfc/bcp/bcp47.txt
  .. _COPYING: COPYING.html

* reStructuredText:

  - Most directives now support a "name" option that attaches a
    reference name.

  - Directive content may start on the first line also when the directive
    type accepts options.

* docs/dev/policies.txt:

  - Recommend the 2-Clause BSD license
    (http://www.spdx.org/licenses/BSD-2-Clause)
    for code that is kept under the author's copyright.

* tools/buildhtml.py:

  - Fix ``--local`` switch.

* Fix [ 3018371 ] Added Lithuanian mappings by Dalius Dobravolskas.

* docutils/writers/html4css1/__init__.py

  - Set "lang" argument for objects with class argument
    "language-<language tag>".
  - New setting "math-output" with support for HTML, MathML, and LaTeX.

* docutils/writers/latex2e/__init__.py

  - Fix [ 3043986 ] AttributeError using :local: with table of content.
  - Place title data in the document preamble.
  - Load `babel` package only if required.
  - Update list of supported languages.
  - New config setting "hyperref-options".
    No hard-coded "unicode" hyperref option (clash with xetex).
  - Set language for custom roles, paragraphs, block-quotes, and
    line-quotes with class argument "language-<language tag>".
  - Fix [ 3095603 ] wrong quotes output for russian and other languages.
  - Convert image URI to a local file path.
  - Apply [ 3148141 ] fix multicolumn support when a colspanning cell
    has more than one paragraph (Wolfgang Scherer).

docutils/docutils/utils/code_analyzer.py
#!/usr/bin/python
# coding: utf-8

"""Lexical analysis of formal languages (i.e. code) using Pygments."""

# :Author: Georg Brandl; Felix Wiemann; Günter Milde
# :Date: $Date: 2011-12-20 15:14:21 +0100 (Di, 20 Dez 2011) $
# :Copyright: This module has been placed in the public domain.

from docutils import ApplicationError
try:
    import pygments
    from pygments.lexers import get_lexer_by_name
    from pygments.formatters.html import _get_ttype_class
    with_pygments = True
except ImportError:
    with_pygments = False

# Filter the following token types from the list of class arguments:
unstyled_tokens = ['token', # Token (base token type)
                   'text',  # Token.Text
                   '']      # short name for Token and Text
# (Add, e.g., Token.Punctuation with ``unstyled_tokens += 'punctuation'``.)

class LexerError(ApplicationError): 
    pass

class Lexer(object):
    """Parse `code` lines and yield "classified" tokens.

    Arguments

      code       -- string of source code to parse,
      language   -- formal language the code is written in,
      tokennames -- either 'long', 'short', or '' (see below).

    Merge subsequent tokens of the same token-type.

    Iterating over an instance yields the tokens as ``(tokentype, value)``
    tuples. The value of `tokennames` configures the naming of the tokentype:

      'long':  downcased full token type name,
      'short': short name defined by pygments.token.STANDARD_TYPES
               (= class argument used in pygments html output),
      'none':      skip lexical analysis.
    """

    def __init__(self, code, language, tokennames='short'):
        """
        Set up a lexical analyzer for `code` in `language`.
        """
        self.code = code
        self.language = language
        self.tokennames = tokennames
        self.lexer = None
        # get lexical analyzer for `language`:
        if language in ('', 'text') or tokennames == 'none':
            return
        if not with_pygments:
            raise LexerError('Cannot analyze code. '
                                    'Pygments package not found.')
        try:
            self.lexer = get_lexer_by_name(self.language)
        except pygments.util.ClassNotFound:
            raise LexerError('Cannot analyze code. '
                'No Pygments lexer found for "%s".' % language)

    # Since version 1.2. (released Jan 01, 2010) Pygments has a
    # TokenMergeFilter. However, this requires Python >= 2.4. When Docutils
    # requires same minimal version,  ``self.merge(tokens)`` in __iter__ can
    # be replaced by ``self.lexer.add_filter('tokenmerge')`` in __init__.
    def merge(self, tokens):
        """Merge subsequent tokens of same token-type.

           Also strip the final newline (added by pygments).
        """
        tokens = iter(tokens)
        (lasttype, lastval) = tokens.next()
        for ttype, value in tokens:
            if ttype is lasttype:
                lastval += value
            else:
                yield(lasttype, lastval)
                (lasttype, lastval) = (ttype, value)
        if lastval.endswith('\n'):
            lastval = lastval[:-1]
        if lastval:
            yield(lasttype, lastval)

    def __iter__(self):
        """Parse self.code and yield "classified" tokens.
        """
        if self.lexer is None:
            yield ([], self.code)
            return
        tokens = pygments.lex(self.code, self.lexer)
        for tokentype, value in self.merge(tokens):
            if self.tokennames == 'long': # long CSS class args
                classes = str(tokentype).lower().split('.')
            else: # short CSS class args
                classes = [_get_ttype_class(tokentype)]
            classes = [cls for cls in classes if cls not in unstyled_tokens]
            yield (classes, value)


class NumberLines(object):
    """Insert linenumber-tokens at the start of every code line.

    Arguments

       tokens    -- iterable of ``(classes, value)`` tuples
       startline -- first line number
       endline   -- last line number

    Iterating over an instance yields the tokens with a
    ``(['ln'], '<the line number>')`` token added for every code line.
    Multi-line tokens are splitted."""

    def __init__(self, tokens, startline, endline):
        self.tokens = tokens
        self.startline = startline
        # pad linenumbers, e.g. endline == 100 -> fmt_str = '%3d '
        self.fmt_str = '%%%dd ' % len(str(endline))

    def __iter__(self):
        lineno = self.startline
        yield (['ln'], self.fmt_str % lineno)
        for ttype, value in self.tokens:
            lines = value.split('\n')
            for line in lines[:-1]:
                yield (ttype, line + '\n')
                lineno += 1
                yield (['ln'], self.fmt_str % lineno)
            yield (ttype, lines[-1])


docutils/docutils/writers/xetex/__init__.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

# :Author: Günter Milde <milde@users.sourceforge.net>
# :Revision: $Revision: 7389 $
# :Date: $Date: 2012-03-30 13:58:21 +0200 (Fr, 30 Mär 2012) $
# :Copyright: © 2010 Günter Milde.
# :License: Released under the terms of the `2-Clause BSD license`_, in short:
# 
#    Copying and distribution of this file, with or without modification,
#    are permitted in any medium without royalty provided the copyright
#    notice and this notice are preserved.
#    This file is offered as-is, without any warranty.
# 
# .. _2-Clause BSD license: http://www.spdx.org/licenses/BSD-2-Clause

"""
XeLaTeX document tree Writer.

A variant of Docutils' standard 'latex2e' writer producing output
suited for processing with XeLaTeX (http://tug.org/xetex/).
"""

__docformat__ = 'reStructuredText'

import os
import os.path
import re

import docutils
from docutils import frontend, nodes, utils, writers, languages
from docutils.writers import latex2e

class Writer(latex2e.Writer):
    """A writer for Unicode-based LaTeX variants (XeTeX, LuaTeX)"""

    supported = ('xetex','xelatex','luatex')
    """Formats this writer supports."""

    default_template = 'xelatex.tex'
    default_preamble = '\n'.join([
        r'% Linux Libertine (free, wide coverage, not only for Linux)',
        r'\setmainfont{Linux Libertine O}',
        r'\setsansfont{Linux Biolinum O}',
        r'\setmonofont[HyphenChar=None]{DejaVu Sans Mono}',
    ])

    config_section = 'xetex writer'
    config_section_dependencies = ('writers', 'latex2e writer')

    settings_spec = frontend.filter_settings_spec(
        latex2e.Writer.settings_spec,
        'font_encoding',
        template=('Template file. Default: "%s".' % default_template,
          ['--template'], {'default': default_template, 'metavar': '<file>'}),
        latex_preamble=('Customization by LaTeX code in the preamble. '
          'Default: select PDF standard fonts (Times, Helvetica, Courier).',
          ['--latex-preamble'],
          {'default': default_preamble}),
        )

    def __init__(self):
        latex2e.Writer.__init__(self)
        self.settings_defaults.update({'fontencoding': ''}) # use default (EU1 or EU2)
        self.translator_class = XeLaTeXTranslator


class Babel(latex2e.Babel):
    """Language specifics for XeTeX.

    Use `polyglossia` instead of `babel` and adapt settings.
    """
    language_codes = latex2e.Babel.language_codes.copy()
    # Additionally supported or differently named languages:
    language_codes.update({
        # code          Polyglossia-name       comment
        'cop':          'coptic',
        'de':           'german', # new spelling (de_1996)
        'de_1901':      'ogerman', # old spelling
        'dv':           'divehi',  # Maldivian
        'dsb':          'lsorbian',
        'el_polyton':   'polygreek',
        'fa':           'farsi',
        'grc':          'ancientgreek',
        'hsb':          'usorbian',
        'sh-cyrl':      'serbian', # Serbo-Croatian, Cyrillic script
        'sh-latn':      'croatian', # Serbo-Croatian, Latin script
        'sq':           'albanian',
        'sr':           'serbian', # Cyrillic script (sr-cyrl)
        'th':           'thai',
        'vi':           'vietnamese',
        # zh-latn:      ???        #     Chinese Pinyin
        })
    # Languages without Polyglossia support:
    for key in ('af',           # 'afrikaans',
                'de_at',        # 'naustrian',
                'de_at_1901',   # 'austrian',
                'fr_ca',        # 'canadien',
                'grc_ibycus',   # 'ibycus', (Greek Ibycus encoding)
                'sr-latn',      # 'serbian script=latin'
               ):
        del(language_codes[key])

    def __init__(self, language_code, reporter):
        self.language_code = language_code
        self.reporter = reporter
        self.language = self.language_name(language_code)
        self.otherlanguages = {}
        self.warn_msg = 'Language "%s" not supported by XeTeX (polyglossia).'
        self.quote_index = 0
        self.quotes = ('"', '"')
        # language dependent configuration:
        # double quotes are "active" in some languages (e.g. German).
        self.literal_double_quote = u'"' # TODO: use \textquotedbl

    def __call__(self):
        setup = [r'\usepackage{polyglossia}',
                 r'\setdefaultlanguage{%s}' % self.language]
        if self.otherlanguages:
            setup.append(r'\setotherlanguages{%s}' %
                         ','.join(self.otherlanguages.keys()))
        return '\n'.join(setup)


class XeLaTeXTranslator(latex2e.LaTeXTranslator):

    def __init__(self, document):
        self.is_xetex = True  # typeset with XeTeX or LuaTeX engine
        latex2e.LaTeXTranslator.__init__(self, document, Babel)
        if self.latex_encoding == 'utf8':
            self.requirements.pop('_inputenc', None)
        else:
            self.requirements['_inputenc'] = (r'\XeTeXinputencoding %s '
                                              % self.latex_encoding)
